lighthouse-core/audits/third-party-facades.js

```diff
@@ -0,0 +1,241 @@
+   * @example {Facebook Messenger Customer Chat} productName
+   */
+  categorySocial: '{productName} (Social)',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+const CATEGORY_UI_MAP = new Map([
+  ['video', UIStrings.categoryVideo],
+  ['customer-success', UIStrings.categoryCustomerSuccess],
+  ['marketing', UIStrings.categoryMarketing],
+  ['social', UIStrings.categorySocial],
+]);
+
+/** @typedef {import("third-party-web").IEntity} ThirdPartyEntity */
+/** @typedef {import("third-party-web").IProduct} ThirdPartyProduct*/
+/** @typedef {import("third-party-web").IFacade} ThirdPartyFacade*/
+
+/** @typedef {{
+ *  product: ThirdPartyProduct,
+ *  cutoffTime: number,
+ *  urlSummaries: Map<string, ThirdPartySummary.Summary>
+ * }} FacadableProductSummary
+ */
+
+class ThirdPartyFacades extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'third-party-facades',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['traces', 'devtoolsLogs', 'URL'],
+    };
+  }
+
+  /**
+   * @param {Map<string, ThirdPartySummary.Summary>} byURL
+   * @param {ThirdPartyEntity | undefined} mainEntity
+   * @return {FacadableProductSummary[]}
+   */
+  static getFacadableProductSummaries(byURL, mainEntity) {
+    /** @type {Map<string, Map<string, FacadableProductSummary>>} */
+    const entitySummaries = new Map();
+
+    for (const [url, urlSummary] of byURL) {
+      const entity = thirdPartyWeb.getEntity(url);
+      if (!entity || thirdPartyWeb.isFirstParty(url, mainEntity)) continue;
+
+      const productSummaries = entitySummaries.get(entity.name) || new Map();
+      const product = thirdPartyWeb.getProduct(url);
+      if (product && product.facades && product.facades.length) {
+        // Record new url if product has a facade.
+
+        /** @type {FacadableProductSummary} */
+        const productSummary = productSummaries.get(product.name) || {
+          product,
+          cutoffTime: Infinity,
+          urlSummaries: new Map(),
+        };
+
+        productSummary.urlSummaries.set(url, urlSummary);
+
+        // This is the time the product resource is fetched.
+        // Any resources of the same entity fetched after this point are considered as part of this product.
+        productSummary.cutoffTime = Math.min(productSummary.cutoffTime, urlSummary.firstEndTime);
+
+        productSummaries.set(product.name, productSummary);
+      }
+      entitySummaries.set(entity.name, productSummaries);
+    }
+
+    for (const [url, urlSummary] of byURL) {
+      const entity = thirdPartyWeb.getEntity(url);
+      if (!entity || thirdPartyWeb.isFirstParty(url, mainEntity)) continue;
+
+      const productSummaries = entitySummaries.get(entity.name) || new Map();
```

Comment:
if there's no existing `productSummaries` this can just `continue`, right? (and no need to re `.set()` it below)